cmd/testbeacon.go

```diff
@@ -181,15 +205,170 @@ func runBeaconTest(ctx context.Context, queuedTestCases []testCaseName, allTestC
 	}
 }
 
-func beaconPing(ctx context.Context, _ *testBeaconConfig, _ string) testResult {
-	// TODO(kalo): implement real ping
-	select {
-	case <-ctx.Done():
-		return testResult{Verdict: testVerdictFail}
-	default:
-		return testResult{
-			Verdict: testVerdictFail,
-			Error:   errNotImplemented,
-		}
+func beaconPingTest(ctx context.Context, _ *testBeaconConfig, target string) testResult {
+	testRes := testResult{Name: "Ping"}
+
+	client := http.Client{}
+	targetEndpoint := fmt.Sprintf("%v/eth/v1/node/health", target)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, targetEndpoint, nil)
+	if err != nil {
+		return failedTestResult(testRes, err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return failedTestResult(testRes, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode > 399 {
+		return failedTestResult(testRes, errors.New("status code %v", z.Int("status_code", resp.StatusCode)))
+	}
+
+	testRes.Verdict = testVerdictOk
+
+	return testRes
+}
+
+func beaconPingMeasureTest(ctx context.Context, _ *testBeaconConfig, target string) testResult {
+	testRes := testResult{Name: "PingMeasure"}
+
+	var start time.Time
+	var firstByte time.Duration
+
+	trace := &httptrace.ClientTrace{
+		GotFirstResponseByte: func() {
+			firstByte = time.Since(start)
+		},
+	}
+
+	start = time.Now()
+	targetEndpoint := fmt.Sprintf("%v/eth/v1/node/health", target)
+	req, err := http.NewRequestWithContext(httptrace.WithClientTrace(ctx, trace), http.MethodGet, targetEndpoint, nil)
+	if err != nil {
+		return failedTestResult(testRes, err)
+	}
+
+	resp, err := http.DefaultTransport.RoundTrip(req)
+	if err != nil {
+		return failedTestResult(testRes, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode > 399 {
+		return failedTestResult(testRes, errors.New("status code %v", z.Int("status_code", resp.StatusCode)))
+	}
+
+	if firstByte > thresholdMeasureBad {
+		testRes.Verdict = testVerdictBad
+	} else if firstByte > thresholdMeasureAvg {
+		testRes.Verdict = testVerdictAvg
+	} else {
+		testRes.Verdict = testVerdictGood
+	}
+	testRes.Measurement = Duration{firstByte}.String()
+
+	return testRes
+}
+
+func beaconIsSyncedTest(ctx context.Context, _ *testBeaconConfig, target string) testResult {
+	testRes := testResult{Name: "isSynced"}
+
+	type isSyncedResponseData struct {
```

Comment:
Mh I had the same thought while reviewing this PR. I think some of the tests in here should be implemented through direct API calls because they need to measure timings -- others, like this one, don't and could be converted to go-eth2-client API calls.

To be fair I'm not very adamant about moving to go-eth2-client here, but we should keep an eye on it and see what goes wrong.

I wonder if we could move the measurements to go-eth2-client as well? Maybe passing some sort of custom HTTP client that does the measurement?

Reply:
First of all, let's not duplicate DTOs, for this case we can simply use [SyncState](https://github.com/attestantio/go-eth2-client/blob/6ff56f91e9dad77f4ff6954d90d8c395123ba38e/api/v1/syncstate.go#L26).

Second point: I think it might be great to test BN connectivity using go-eth2-client intentionally, because if something is wrong due to the client library, we will catch it proactively.

As for measurements, I don't see how "direct call" approach would benefit here.. after all, the client does all the same, maybe using different default timings, which we can override by implementing our custom [Service](https://github.com/attestantio/go-eth2-client/blob/master/http/service.go#L38).

Reply:
Yes, my idea was not to rely on external library and catch if there are any underlying changes. Completely fine by me to change

```
	type isSyncedResponse struct {
		Data isSyncedResponseData `json:"data"`
	}
```

to

```
	type isSyncedResponse struct {
		Data eth2v1.SyncState `json:"data"`
	}
```

and delete `isSyncedResponseData`.